containerfixture/containerfixture.go

```
@@ -0,0 +1,119 @@
+var manifest string
+
+const Username = "admin"
+const Password = "Password"
+
+// Generates a random dynamic port number
+func GetRandomPort() string {
+	minPort := 49152
+	maxPort := 65535
+	rand.Seed(time.Now().UnixNano())
+
+	randomPort := fmt.Sprintf("%d", rand.Intn(maxPort-minPort)+minPort)
+	return randomPort
+}
+
+// Modify the manifest to allow running it in parallel as
+// many times as needed with dynamic values
+func modifyYaml(yaml string) map[string]string {
+	podName := fmt.Sprintf("test-pod-%s", GetRandomPort())
+	webPort := GetRandomPort()
+	apiPort := GetRandomPort()
+
+	yaml = strings.Replace(yaml, "name: \"test-pod\"", fmt.Sprintf("name: %s", podName), 1)
+	yaml = strings.Replace(yaml, "hostPort: 8000", fmt.Sprintf("hostPort: %s", webPort), 1)
+	yaml = strings.Replace(yaml, "hostPort: 8089", fmt.Sprintf("hostPort: %s", apiPort), 1)
+
+	return map[string]string{
+		"yaml":    yaml,
+		"podName": podName,
+		"webPort": webPort,
+		"apiPort": apiPort,
+	}
+}
+
+// Building and Running a k8s manifest using 'podman kube play' command.
+// It will make up to three attempts to run with different configurations,
+// allowing it to run concurrently if necessary
+func BuildAndRunPod() map[string]string {
+	var deployment map[string]string
+	retries := 3
+	for retry := 0; retry < retries; retry++ {
+		deployment := modifyYaml(manifest)
+
+		cmd := exec.Command("podman", "play", "kube", "-")
+		cmd.Stdin = strings.NewReader(deployment["yaml"])
+		output, err := cmd.CombinedOutput()
+		if err == nil {
+			outputString := strings.Split(strings.TrimSpace(string(output)), "\n")
+			containerID := outputString[len(outputString)-1]
+			deployment["containerID"] = containerID
+			return deployment
+		}
+		log.Printf("Failed to start pod (attempt %d of %d).\n", retry+1, retries)
+	}
+	Cleanup("", deployment["podName"])
+	log.Fatalf("Failed to start pod after %d retries", retries)
+	return make(map[string]string) // Avoid 'missing return' error
```

Why not return an error?

---

`log.Fatalf` already  call to os.Exit(1), the `return.make` will never be reached (but GO requires it). how would you suggest to change it?

---

This should be failing the test not the whole testing process.

---

fixed, replaced with `t.Fail()` alongside `t.Logf`